sim/midas/src/main/scala/midas/widgets/ClockBridge.scala

```diff
@@ -143,13 +143,38 @@ class ClockTokenVector(numClocks: Int) extends Bundle with HasChannels with Cloc
 class ClockBridgeModule(params: ClockParameters)(implicit p: Parameters)
     extends BridgeModule[ClockTokenVector] {
   val clockInfo = params.clocks
+  val phaseRelationships = clockInfo map { cInfo => (cInfo.multiplier, cInfo.divisor) }
+
   lazy val module = new BridgeModuleImp(this) {
+
   val io = IO(new WidgetIO())
   val hPort = IO(new ClockTokenVector(clockInfo.size))
-  val phaseRelationships = clockInfo map { cInfo => (cInfo.multiplier, cInfo.divisor) }
+  val clocks = hPort.clocks
+
+  // Register used to set the credit.
+  val creditIn = Wire(Decoupled(UInt(32.W)))
+  genAndAttachQueue(creditIn, "CREDIT")
+  creditIn.ready := true.B
+
+  // Register which reports the number of remaining tokens to the driver.
+  val credit = RegInit(0.U(32.W))
+  attach(credit, "CREDIT_COUNT", ReadOnly)
+
+  // Reset the credit count if the register is written. Otherwise, whenever a
+  // clock token is delivered, decrement the count. Stop when it reaches 0.
+  when (creditIn.fire) {
+    credit := creditIn.bits
+  }.elsewhen (clocks.fire) {
+    credit := credit - 1.U
+  }
```

Comment:
The problem her right now is that from the software side you can do `credit(100); credit(100);` and you'd expect there to be 200 credits in total arriving at the hardware. In practice there's likely going to be a data race here where the first call sets the credit to 100, which starts ticking down, but before it's at 0 the second call will reset the credit counter to 100, overwriting the non-zero count in there.

I see two approaches to fix this: One would be to add the new credits to the counter instead of resetting the counter, and also lowering the ready wire if the new credit would overflow the counter. That way there's no race  condition and always the exact number of credits coming from the host will be applied.
```
creditIn.ready := false.B
val changedCredit = Wire(credit)
when (creditIn.valid) {
  val addedCredit = newCredit +& creditIn.bits
  when (addedCredit >> 32 == 0.U) {
    changedCredit := addedCredit
    creditIn.ready := true.B
  }
}
when (clocks.fire) {
  changedCredit := changedCredit - 1.U
}
credit := changedCredit
```
An alternative approach would be to just refuse to accept new credit until the old has been drained:
```
creditIn.ready := false.B
val changedCredit = Wire(credit)
when (creditIn.valid && changedCredit == 0) {
  changedCredit := creditIn.bits
  creditIn.ready = true.B
}
when (clocks.fire) {
  changedCredit := changedCredit - 1.U
}
credit := changedCredit
```

In both cases I think you'll want to have the effects of `creditIn` and `clocks` on the credit count stack up, such that credit doesn't get lost. With the `elsewhen` approach you wouldn't subtract 1 from the credit on a clock token if there's a creditIn at the same time.

Reply:
I agree, incrementing the credit count makes sense. Updated the implementation.